2일차 미션 - 문자열 계산기

lazylune opened the pull request:
업무로 인하여 많이 늦은 점 양해 부탁 드립니다.
앞으로 꾸준히 시간 날 때마다 구현하여 풀리퀘 드리겠습니다.
잘 부탁 드려요.

WoorakChoi committed:
refactor 1단계 미션 코드 리팩토링

 - package 명 study -> basic 으로 변경
 - 테스트 대상이 되는 코드는 main 디렉토리 이하로 이동

WoorakChoi committed:
feat 2단계 미션 문자열 계산기

 - String?.tokenize(): 문자열을 공백 문자 기준 token 으로 분해한 리스트를 리턴한다.
 - List<String>.parse(): 사칙연산 계산을 위한 Expression 트리를 생성한 뒤 리턴한다.
 - String?.calculate(): 문자열로 입력된 수식을 계산한 뒤 결과 값을 리턴한다. tokenize(), parse() 를 통해 구현

ddaaac commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,71 @@
+            }
+        }
+    }
+}
```
[POSIX new line](https://blog.coderifleman.com/2015/04/04/text-files-end-with-a-newline/)을 이미 알고 계실 수 있을텐데요, [인텔리제이에서는 옵션으로 항상 개행](https://velog.io/@d-h-k/intellij-%ED%8C%8C%EC%9D%BC%EB%81%9D%EC%97%90-%EA%B0%9C%ED%96%89%EC%9D%84-%EC%9E%90%EB%8F%99%EC%9C%BC%EB%A1%9C-%EC%B6%94%EA%B0%80%ED%95%98%EB%8A%94-%EB%B0%A9%EB%B2%95)을 붙여줄 수 있습니다 :)

lazylune replied:
옵션 활성화 시켰습니다. :)

ddaaac commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,71 @@
+package calculator
+
+fun String?.tokenize(): List<String> = this?.trim()?.split("\\s+".toRegex())?.filter { it != "" }
+    ?: throw IllegalArgumentException("Null cannot be tokenized")
```
null이 들어올 경우에는 항상 예외가 발생하고 있어요.
자바에서는 타입에 대한 non-nullable을 보장할 때 항상 런타임에 null 검사를 통해 처리해야했습니다.
반면에, 코틀린은 컴파일 타임에 non-nullable을 보장할 수 있습니다. 
후자 쪽이 좀 더 안전한 프로그래밍이 가능할 것 같은데 우락님의 생각은 어떠신가요?

lazylune replied:
네, 맞습니다. non-nullable 이 훨씬 안전한 프로그래밍이 가능하죠. 여기에서는 "입력값이 null이거나 빈 공백 문자일 경우 IllegalArgumentException throw" 요건을 만족시키기 위해 의도적으로 null 입력을 가능케 하고 예외를 발생시키도록 구현하였습니다.

ddaaac replied:
아마 앞으로 미션을 구현할 때 MVC 패턴을 많이 사용하게 될 것 같아요.
null에 대한 검사는 View에서 하는 것도 적당할 것 같네요 :)

ddaaac commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,71 @@
+    ?: throw IllegalArgumentException("Null cannot be tokenized")
+
+fun List<String>.parse(): Expression {
+    fun parse(exp: Expression, tokenList: List<String>): Expression {
```
```suggestion
    tailrec fun parse(exp: Expression, tokenList: List<String>): Expression {
```
코틀린은 [꼬리 재귀 최적화](https://kotlinlang.org/docs/functions.html#tail-recursive-functions)를 지원합니다 😮 

실제로 어떻게 컴파일 되는지는 [교육 자료](https://edu.nextstep.camp/s/MFvxefXA/ls/rclWyBPz)에서 확인하실 수 있습니다 😄

lazylune replied:
재귀 함수를 실제 코드에 적용하기 시작한지는 얼마 되지 않아서..
코드를 짜면서도 꼬리 재귀인지 아닌지 판단할 새도 없이 일단 구현부터 하고 봤네요. :)
tailrec 꼭 기억하겠습니다.

ddaaac commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,71 @@
+        }
+    }
+
+    data class Operator(
+        val function: (Int, Int) -> Int,
+        var firstOperand: Expression = Undefined,
+        var secondOperand: Expression = Undefined
+    ) : Expression() {
```
요 BinaryExpression은 양 옆의 Expression이 채워지기 전에는 불완전한 객체가 될 거 같아요.
파싱하면서 온전한 Expression 객체를 만들 수 있는 방법이 있나 찾아봤더니 [이런 방법](https://craftinginterpreters.com/parsing-expressions.html#recursive-descent-parsing)이 있더라구요.
이 부분은 리뷰 반영하시지는 않아도 되구, 나중에 흥미가 생기시면 도전해보면 좋을 것 같습니다 😄

lazylune replied:
Full Parser 군요! 내용이 길어 나중에 각잡고 읽어봐야 할 듯 싶네요. :) 좋은 자료 감사합니다. 👍  원래는 불완전 식을 파싱 과정에서 잡아야하는게 정상인데. 방법이 생각나지 않아 꼼수로 Undefined 를 추가하고 eval 과정에서 잡히도록 했네요. 나중에 개선해 보도록 하겠습니다.

ddaaac commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,71 @@
+    fun parse(exp: Expression, tokenList: List<String>): Expression {
+        if (tokenList.isEmpty()) return exp
+
+        val newExp = when (val parsed = Expression.parse(tokenList.first())) {
```
sealed class 와 when 절의 조합 좋네요 👍 

-----

이 부분에서 when 절이 중첩되면서 depth가 깊어지고 있어요.
미션을 진행하는 동안에는 메서드 분리를 통해 의식적으로 depth가 1을 넘지 않게 코딩해보는 것도 좋은 경험일 것 같습니다!

lazylune replied:
메서드의 적당한 이름이 생각이 안나, 여기에서는 이대로 유지했으면 합니다. 아니면 좋은 이름 추천해 주셔도 좋아요^^

ddaaac replied:
제가 네이밍은 좀 많이 약한 편이라서요... 😅 참고만해주세요!
기존 expression에 새로운 expression을 적용하는 것이니 apply 같은 메서드 명도 괜찮을 것 같기도 해요.
그렇게 쓰고보니깐 로직이 Expression 내부로 이사가도 괜찮을 것 같다는 생각이 들더라구요.

```kotlin
            val newExp = apply(exp, Expression.parse(tokenList.first()))
            val newExp = exp.apply(Expression.parse(tokenList.first())))
```

ddaaac commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,71 @@
+        val newExp = when (val parsed = Expression.parse(tokenList.first())) {
+            is Expression.Operand -> when (exp) {
+                Expression.Undefined -> parsed
+                is Expression.Operator -> exp.also { it.secondOperand = parsed }
```
어떤 스코프 함수를 사용해야될지 고민될 때가 있는데 저는 [이 글](https://medium.com/@limgyumin/%EC%BD%94%ED%8B%80%EB%A6%B0-%EC%9D%98-apply-with-let-also-run-%EC%9D%80-%EC%96%B8%EC%A0%9C-%EC%82%AC%EC%9A%A9%ED%95%98%EB%8A%94%EA%B0%80-4a517292df29)을 참고하고는 합니다.
위 글에 따르면

- apply
> 수신 객체 의 프로퍼티 만을 사용하는 대표적인 경우가 객체의 초기화 이며

- also
> 객체의 사이드 이팩트를 확인하거나 수신 객체의 프로퍼티에 데이터를 할당하기 전에 해당 데이터의 유효성을 검사 할 때 매우 유용합니다.

라고 하는데요, 이 경우엔 apply가 조금 더 적당하지 않을까 싶어요.

lazylune replied:
when 절 이하의 주된 목적은 newExp 가 무엇인지 결정하는 것으로 보았고, 바꾸는 과정에서의 데이터 초기화를 부수적인 과정이라 보고 also 를 사용했네요. 초기화 관점에서 본다면 apply 도 타당하네요.

ddaaac commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,71 @@
+                "-" -> Operator(Int::minus)
+                "*" -> Operator(Int::times)
+                "/" -> Operator(Int::div)
+                else -> try {
+                    Operand(input.toInt())
+                } catch (e: NumberFormatException) {
+                    throw IllegalArgumentException(e)
+                }
```
```suggestion
                else -> {
                    val parsed = input.toIntOrNull() ?: throw IllegalArgumentException()
                    Operand(parsed)
                }
```
이 부분은 취향 차겠지만 앞으로 `xxxOrNull()` + 엘비스 연산자(?:)를 사용한 예외처리를 많이 보시게 될 것 같아서 남기는 코멘트입니다. 코틀린의 많은 API들이 `xxxOrNull`과 같은 네이밍으로 예외가 발생했을 때 null을 리턴하는데요, 커스텀 예외를 발생시키고 싶다면 자주 사용하게 되는 것 같아요 😄

lazylune replied:
try 보다 엘비스가 깔끔하네요.
input.toIntOrNull()?.let { Operand(it) } ?: throw IllegalArgumentException()
로 parsed 로컬 변수까지 없애는건 너무 나간걸까요? ;)

ddaaac replied:
적절하게 `let`을 사용한 예시인 것 같아요 👍 

체이닝으로 표현하는 것과 적절하게 변수로 끊어주며 코딩하는 방식은 취향에 따라 적절하게 구현해주시면 될 것 같습니다!
저 같은 경우는 예전에는 체이닝을 선호하다가 최근에는 적당히 끊어가며 작성하는 편이에요 😄

ddaaac commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,71 @@
+
+fun String?.calculate(): Int = this.tokenize().parse().eval()
+
+sealed class Expression {
```
kotlin 1.5부터는 sealed interface를 지원합니다. interface를 사용해도 좋을 것 같아요 :)

lazylune replied:
오오~! sealed interface 좋네요.

ddaaac commented on src/test/kotlin/calculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,176 @@
+
+class TokenizeTest {
+
+    @ParameterizedTest
+    @ValueSource(strings = ["", " ", "   ", "\t", "\n", "\t\n", "\n\t", " \n \t"])
+    fun `공백 문자를 입력하면 빈 리스트를 출력한다`(input: String) {
+        assertThat(input.tokenize()).isEmpty()
+    }
+
+    @Test
+    fun `앞선 공백 문자나 연속된 공백 문자, 뒤의 공백문자는 제거된다`() {
+        assertThat("  1 + 2".tokenize()).isEqualTo(listOf("1", "+", "2"))
+        assertThat("13 - 21   ".tokenize()).isEqualTo(listOf("13", "-", "21"))
+        assertThat("15543 * 21121 \n ".tokenize()).isEqualTo(listOf("15543", "*", "21121"))
+        assertThat("-231 \t  /  232".tokenize()).isEqualTo(listOf("-231", "/", "232"))
+    }
```
꼼꼼하게 테스트를 잘해주셨어요 👍 
아래 테스트도 ParameterizedTest로 만들 수 있을 것 같습니다!

lazylune replied:
MethodSource 로 변경하였습니다. List 도 CsvSource 가 되는지 모르겠네요.

ddaaac commented on src/test/kotlin/calculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,176 @@
+
+class CalculatorTest {
+
+    @ParameterizedTest
+    @MethodSource("providePlusTestArgument")
+    fun `덧셈 연산`(input: String, result: Int) {
+        assertThat(input.calculate()).isEqualTo(result)
+    }
```
여기 데이터는 `MethodSource`로도 공급할 수 있지만 `CsvSource`로도 공급할 수 있을 거 같아요.
저는 `CsvSource`를 사용하는 게 가능하다면 `CsvSource`를 사용하는 게 더 테스트 코드를 읽기 좋다고 생각해요.
아무래도 `MethodSource`를 사용하게 되면 물리적으로 테스트 코드와 데이터 사이에 거리가 생기고, 읽으려면 데이터와 테스트 사이를 왔다갔다 해야되더라구요. 
우락님은 어떻게 생각하시나요?

lazylune replied:
CsvSource 는 유심히 안봤는데.. 다시 한번 살펴봐야겠네요.

lazylune replied:
String 으로 명시하고 있어 int 나 boolean 은 어떻게 파라미터로 받을까 의문이었는데. 자동으로 형변환이 되는 거였군요~! input output 이 primitive type 이라면 말씀하신대로 CsvSource 를 선호할 것 같습니다.

ddaaac commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,71 @@
+fun String?.tokenize(): List<String> = this?.trim()?.split("\\s+".toRegex())?.filter { it != "" }
+    ?: throw IllegalArgumentException("Null cannot be tokenized")
+
+fun List<String>.parse(): Expression {
```
parse를 하는 객체를 분리해도 괜찮을 것 같습니다 😄

lazylune replied:
객체 설계 이야기가 나와서 문의 드리는데요. 처음에는 객체로 설계를 했다가 tokenize() 나 parse() 가 private 함수로 구현해야 될 것 같은데, 그럼 테스트가 여의치 않아 보여 위와 같이 확장 함수로 구현했습니다. private 함수 테스트는 보통 어떻게 진행하시나요?

ddaaac replied:
private 함수를 테스트하고 싶을 때는 리팩토링이 필요하지 않을까 생각해봅니다. [이 스레드](https://stackoverflow.com/a/105021)를 한 번 보면 좋을 것 같아요. 
> If I find that the private method is huge or complex or important enough to require its own tests, I just put it in another class and make it public there.

lazylune replied:
좋은 의견입니다. 객체로 분리하여 parse 부분에 대한 유연성을 높이는 방향으로 설계를 변경하였습니다.

ddaaac reviewed (changes requested):
안녕하세요 ✋ 

바쁘신 와중에도 미션 정말 잘 구현해주셨습니다 💯 

몇 가지 코멘트를 남겼으니 확인 부탁드려요!

WoorakChoi committed:
style 파일 끝에 개행 추가

WoorakChoi committed:
fix parse 코드에 꼬리 재귀 최적화 적용

WoorakChoi committed:
refactor Expression sealed class 를 sealed interface 로 변경

WoorakChoi committed:
fix 숫자 문자열 변환 시 try-catch 대신 toIntOrNull() 사용

WoorakChoi committed:
fix ParameterizedTest 방식 변경: MethodSource -> CsvSource

WoorakChoi committed:
refactor Parser 및 Calculator 객체화

 - Parser 인터페이스 추가
 - 우선 순위를 고려하지 않는 파서 (Non-precedence parser) 클래스에 parse 함수 이동
 - Calculator 는 parser 를 외부에서 주입받아 추후 우선 순위를 고려한 parser 도 쉽게 대체할 수 있도록 설계 변경

lazylune commented:
리뷰해 주신 부분 수정하여 commit 하였습니다.

ddaaac commented on src/main/kotlin/calculator/Calculator.kt:
```diff
@@ -0,0 +1,77 @@
+    }
+}
+
+class Calculator(private val parser: Parser = NonPrecedenceParser()) {
+
+    fun calculate(input: String): Int = parser.parse(input.tokenize()).eval()
+}
+
+sealed interface Expression {
```
여러 클래스를 한 파일에 모두 두셨는데 특별한 이유가 있으실까요?

저는 Parser/Expression/Calculator(<- 얘는 조금 애매하네요) 정도로 나눌 수 있을 것 같아요.
저도 인터페이스 - 구현체는 한 파일에 두는 경우가 자주 있는데(특히 sealed 클래스일 경우에는요) 계산기 관련된 클래스들을 모두 한 파일에 두면 나중에 찾기 조금 힘들지 않을까 싶은 생각이 드네요.

ddaaac reviewed (approved):
안녕하세요 👋 

리뷰 반영 너무 잘해주셨습니다 💯 
코멘트 꼼꼼하게 남겨주셔서 저도 많이 배워가네요.

몇 가지 코멘트와 답변 남겼으니 확인 부탁드려요!

다음 단계에 같이 반영해주심 될 것 같습니다~
